File: DrawRobotControl/SAFReader/SAFCrypto.cpp
#include "stdafx.h"

const unsigned char* GetSAFEncryptionKey()
{
	static const unsigned char gEncryptionKey[32] = { 0xAE,0xEC,0x5E,0x77,0x92,0xFF,0x0D,0x37,0xE0,0xA3,0x82,0x77,0x20,0x5A,0xC4,0xCC,0xE7,0x83,0xD8,0xE9,0x49,0x8D,0x87,0x38,0x00,0xB5,0x9B,0xE9,0x5C,0x1B,0xE5,0x7F };
	return gEncryptionKey;
}

void Decrypt1Block_CBC_AES_256_NOAUTH(const byte* iv, const byte* data, size_t dataLen, byte* out_data)
{
	// Create the decryption object
	CryptoPP::CBC_Mode<CryptoPP::AES>::Decryption decryptor;
	decryptor.SetKeyWithIV(GetSAFEncryptionKey(), CryptoPP::AES::MAX_KEYLENGTH, iv);
	decryptor.ProcessData(out_data, data, dataLen);
}

void GetDataBlockHash(byte* iv, byte* data, size_t dataSize, byte* hash)
{
	CryptoPP::HMAC<CryptoPP::SHA256> hmac(GetSAFEncryptionKey(), CryptoPP::AES::MAX_KEYLENGTH);
	hmac.Update(iv, CryptoPP::AES::BLOCKSIZE); // Hash the first block of data
	hmac.Update(data, dataSize); // Hash the second block of data
	hmac.Final(hash); // Finalize the hash
}

int ReadGenericEncryptedBlock(FILE* f, size_t blockSize, byte* out_dec)
{
	byte iv[CryptoPP::AES::BLOCKSIZE];
	byte hash[CryptoPP::SHA256::DIGESTSIZE];
	byte hash_check[CryptoPP::SHA256::DIGESTSIZE];

	size_t readCount;
	readCount = fread(iv, 1, CryptoPP::AES::BLOCKSIZE, f);
	if (readCount != CryptoPP::AES::BLOCKSIZE)
	{
		printf("ERROR : File does not contain enough data to read IV\n");
		return -1;
	}

	byte* tempEncData = (byte*)malloc(blockSize);
	if (tempEncData == NULL)
	{
		printf("ERROR : Failed to allocate %zd bytes of data\n", blockSize);
		return -2;
	}

	readCount = fread(tempEncData, 1, blockSize, f);
	if (readCount != blockSize)
	{
		printf("File does not contain enough data to read block data\n");
		free(tempEncData);
		return -2;
	}

	readCount = fread(hash, 1, CryptoPP::SHA256::DIGESTSIZE, f);
	if (readCount != CryptoPP::SHA256::DIGESTSIZE)
	{
		printf("File does not contain enough data to read hash data\n");
		free(tempEncData);
		return -2;
	}

	Decrypt1Block_CBC_AES_256_NOAUTH(iv, (byte*)tempEncData, (int)blockSize, (byte*)out_dec);

	GetDataBlockHash(iv, (byte*)tempEncData, (int)blockSize, hash_check);
	if (memcmp(hash, hash_check, CryptoPP::SHA256::DIGESTSIZE) != 0)
	{
		printf("Data authentication failed for block sized %zd\n", blockSize);
	}

	free(tempEncData);

	return 0;
}

int ReadGenericBlock(FILE* f, size_t blockSize, byte* out_dec)
{
	byte hash[CryptoPP::SHA256::DIGESTSIZE];
	byte hash_check[CryptoPP::SHA256::DIGESTSIZE];

	size_t readCount;

	readCount = fread(out_dec, 1, blockSize, f);
	if (readCount != blockSize)
	{
		printf("File does not contain enough data to read block data\n");
		return -2;
	}

	readCount = fread(hash, 1, CryptoPP::SHA256::DIGESTSIZE, f);
	if (readCount != CryptoPP::SHA256::DIGESTSIZE)
	{
		printf("File does not contain enough data to read hash data\n");
		return -2;
	}

	CryptoPP::SHA256 sha256;
	sha256.Update((byte*)out_dec, blockSize); // Hash the first block of data
	sha256.Final(hash_check); // Finalize the hash 

	if (memcmp(hash, hash_check, CryptoPP::SHA256::DIGESTSIZE) != 0)
	{
		printf("Data authentication failed for block sized %zd\n", blockSize);
	}

	return 0;
}

int GetIVSize()
{
	return CryptoPP::AES::BLOCKSIZE;
}

int GetHashSize()
{
	return CryptoPP::SHA256::DIGESTSIZE;
}

void WriteGenericEncryptedBlock(FILE* f, byte* buff, size_t bufferSize)
{
	byte iv[CryptoPP::AES::BLOCKSIZE];
	AutoSeededRandomPool rng;
	rng.GenerateBlock(iv, sizeof(iv));

	CryptoPP::AES::Encryption aesEncryption(GetSAFEncryptionKey(), CryptoPP::AES::MAX_KEYLENGTH);
	CryptoPP::CBC_Mode_ExternalCipher::Encryption cbcEncryption(aesEncryption, iv);

	byte* ciphertext = new byte[bufferSize];
	CryptoPP::ArraySink arraySink(ciphertext, bufferSize); // write to a byte buffer

	CryptoPP::StreamTransformationFilter filter(cbcEncryption);
	filter.Attach(new CryptoPP::Redirector(arraySink)); // attach the ArraySink to write to the buffer
	filter.Put(buff, bufferSize);
	filter.MessageEnd();

	byte hash[CryptoPP::SHA256::DIGESTSIZE];
	GetDataBlockHash(iv, ciphertext, bufferSize, hash);

	fwrite(iv, 1, sizeof(iv), f);
	fwrite(ciphertext, 1, bufferSize, f);
	fwrite(hash, 1, sizeof(hash), f);

	delete[] ciphertext;
}

void WriteGenericBlock(FILE* f, byte* buff, size_t bufferSize)
{
	byte hash[CryptoPP::SHA256::DIGESTSIZE];
	CryptoPP::SHA256 sha256;
	sha256.Update(buff, bufferSize); // Hash the first block of data
	sha256.Final(hash); // Finalize the hash 

	fwrite(buff, 1, bufferSize, f);
	fwrite(hash, 1, sizeof(hash), f);
}

File: DrawRobotControl/SAFFileHandler/SAFCrypto.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System;
using System.Security.Cryptography;

namespace SAFFileHandler
{
    public static class SAFCrypto
    {
        static byte[] SAFEncryptionKey = new byte[] { 0xAE, 0xEC, 0x5E, 0x77, 0x92, 0xFF, 0x0D, 0x37, 0xE0, 0xA3, 0x82, 0x77, 0x20, 0x5A, 0xC4, 0xCC, 0xE7, 0x83, 0xD8, 0xE9, 0x49, 0x8D, 0x87, 0x38, 0x00, 0xB5, 0x9B, 0xE9, 0x5C, 0x1B, 0xE5, 0x7F };
        const int CryptoPP_AES_BLOCKSIZE = 16;
        const int CryptoPP_SHA256_DIGESTSIZE = 32;
        public static void Decrypt1Block_CBC_AES_256_NOAUTH(byte[] iv, byte[] data, int dataLen, byte[] out_data)
        {
            using (Aes aes = Aes.Create())
            {
                aes.Mode = CipherMode.CBC;
                aes.KeySize = 256;
                aes.BlockSize = 128;
                aes.Padding = PaddingMode.None;
                aes.Key = SAFEncryptionKey;
                aes.IV = iv;

                using (ICryptoTransform decryptor = aes.CreateDecryptor())
                {
                    decryptor.TransformBlock(data, 0, dataLen, out_data, 0);
                }
            }
        }
        public static void GetDataBlockHash(byte[] iv, byte[] data, int dataSize, byte[] hash)
        {
            using (HMACSHA256 hmac = new HMACSHA256(SAFEncryptionKey))
            {
                hmac.TransformBlock(iv, 0, iv.Length, iv, 0); // Hash the first block of data
                hmac.TransformBlock(data, 0, dataSize, data, 0); // Hash the second block of data
                hmac.TransformFinalBlock(new byte[0], 0, 0); // Finalize the hash
                if (hmac.Hash != null)
                {
                    Array.Copy(hmac.Hash, hash, hash.Length); // Copy the result to the output array
                }
            }
        }

        public static int ReadGenericEncryptedBlock(FileStream f, int blockSize, byte[] out_dec)
        {
            byte[] iv = new byte[CryptoPP_AES_BLOCKSIZE];
            byte[] hash = new byte[CryptoPP_SHA256_DIGESTSIZE];
            byte[] hash_check = new byte[CryptoPP_SHA256_DIGESTSIZE];

            int readCount;
            readCount = f.Read(iv, 0, CryptoPP_AES_BLOCKSIZE);
            if (readCount != CryptoPP_AES_BLOCKSIZE)
            {
                Console.WriteLine("ERROR: File does not contain enough data to read IV");
                return -1;
            }

            byte[] tempEncData = new byte[blockSize];
            readCount = f.Read(tempEncData, 0, blockSize);
            if (readCount != blockSize)
            {
                Console.WriteLine("File does not contain enough data to read block data");
                return -2;
            }

            readCount = f.Read(hash, 0, CryptoPP_SHA256_DIGESTSIZE);
            if (readCount != CryptoPP_SHA256_DIGESTSIZE)
            {
                Console.WriteLine("File does not contain enough data to read hash data");
                return -2;
            }

            Decrypt1Block_CBC_AES_256_NOAUTH(iv, tempEncData, blockSize, out_dec);

            GetDataBlockHash(iv, tempEncData, blockSize, hash_check);
            if (!hash.SequenceEqual(hash_check))
            {
                Console.WriteLine($"Data authentication failed for block sized {blockSize}");
            }

            return 0;
        }

        public static int ReadGenericBlock(FileStream f, int blockSize, byte[] out_dec)
        {
            byte[] hash = new byte[CryptoPP_SHA256_DIGESTSIZE];
            byte[] hash_check = new byte[CryptoPP_SHA256_DIGESTSIZE];

            int readCount = f.Read(out_dec, 0, blockSize);
            if (readCount != blockSize)
            {
                Console.WriteLine("File does not contain enough data to read block data");
                return -2;
            }

            readCount = f.Read(hash, 0, hash.Length);
            if (readCount != hash.Length)
            {
                Console.WriteLine("File does not contain enough data to read hash data");
                return -2;
            }

            using (SHA256 sha256 = SHA256.Create())
            {
                byte[] hashBytes = sha256.ComputeHash(out_dec, 0, blockSize); // Hash the first block of data
                Buffer.BlockCopy(hashBytes, 0, hash_check, 0, hashBytes.Length); // copy the hash value to hash_check
            }

            if (!hash.SequenceEqual(hash_check))
            {
                Console.WriteLine("Data authentication failed for block sized {0}", blockSize);
            }

            return 0;
        }

        public static void GenerateRandomBytes(byte[] buffer)
        {
            using (var rng = RandomNumberGenerator.Create())
            {
                rng.GetBytes(buffer);
            }
        }

        public static void Encrypt1Block_CBC_AES_256_NOAUTH(byte[] iv, byte[] data, int dataLength, byte[] out_data)
        {
            using (var aes = Aes.Create())
            {
                aes.Mode = CipherMode.CBC;
                aes.Padding = PaddingMode.PKCS7;
                aes.KeySize = 256;
                aes.BlockSize = 128;
                aes.Key = SAFEncryptionKey;

                using (var encryptor = aes.CreateEncryptor(aes.Key, iv))
                {
                    encryptor.TransformBlock(data, 0, dataLength, out_data, 0);
                }
            }
        }

        public static void WriteGenericEncryptedBlock(FileStream f, byte[] buff, int bufferSize)
        {
            byte[] iv = new byte[CryptoPP_AES_BLOCKSIZE];
            GenerateRandomBytes(iv);
            byte[] ciphertext = new byte[bufferSize];
            Encrypt1Block_CBC_AES_256_NOAUTH(iv, buff, bufferSize, ciphertext);

            byte[] hash = new byte[CryptoPP_SHA256_DIGESTSIZE];
            GetDataBlockHash(iv, ciphertext, ciphertext.Length, hash);

            f.Write(iv, 0, iv.Length);
            f.Write(ciphertext, 0, ciphertext.Length);
            f.Write(hash, 0, hash.Length);
        }

        public static void WriteGenericBlock(FileStream f, byte[] buff, int bufferSize)
        {
            byte[]? hash = null;
            using (var sha256 = SHA256.Create())
            {
                hash = sha256.ComputeHash(buff, 0, bufferSize);
            }

            f.Write(buff, 0, bufferSize);
            f.Write(hash, 0, hash.Length);
        }
    }
}
